File: data-structures/binary_tree.py
"""
Binary Tree Implementation in Python

Time Complexity:
- Insert: O(log n) average, O(n) worst case
- Search: O(log n) average, O(n) worst case
- Delete: O(log n) average, O(n) worst case

Space Complexity: O(n) for storage, O(log n) for recursion stack

Python features:
- Object-oriented design
- Generator expressions
- Magic methods
- Type hints
"""

from typing import Optional, List, Iterator, Any
from collections import deque
import json


class TreeNode:
    """A single node in the binary tree."""

    def __init__(self, data: Any) -> None:
        self.data = data
        self.left: Optional["TreeNode"] = None
        self.right: Optional["TreeNode"] = None

    def __str__(self) -> str:
        return str(self.data)

    def __repr__(self) -> str:
        return f"TreeNode({self.data})"


class BinaryTree:
    """
    Binary Tree implementation with various traversal and utility methods.
    """

    def __init__(self) -> None:
        self.root: Optional[TreeNode] = None
        self.size = 0

    def insert(self, data: Any) -> None:
        """Insert a new node into the tree."""
        if self.root is None:
            self.root = TreeNode(data)
        else:
            self._insert_recursive(self.root, data)
        self.size += 1

    def _insert_recursive(self, node: TreeNode, data: Any) -> None:
        """Helper method for recursive insertion."""
        if data < node.data:
            if node.left is None:
                node.left = TreeNode(data)
            else:
                self._insert_recursive(node.left, data)
        else:
            if node.right is None:
                node.right = TreeNode(data)
            else:
                self._insert_recursive(node.right, data)

    def search(self, data: Any) -> bool:
        """Search for a value in the tree."""
        return self._search_recursive(self.root, data)

    def _search_recursive(self, node: Optional[TreeNode], data: Any) -> bool:
        """Helper method for recursive search."""
        if node is None:
            return False

        if data == node.data:
            return True
        elif data < node.data:
            return self._search_recursive(node.left, data)
        else:
            return self._search_recursive(node.right, data)

    def delete(self, data: Any) -> bool:
        """Delete a node from the tree."""
        if self.root is None:
            return False

        self.root, deleted = self._delete_recursive(self.root, data)
        if deleted:
            self.size -= 1
        return deleted

    def _delete_recursive(
        self, node: Optional[TreeNode], data: Any
    ) -> tuple[Optional[TreeNode], bool]:
        """Helper method for recursive deletion."""
        if node is None:
            return None, False

        if data < node.data:
            node.left, deleted = self._delete_recursive(node.left, data)
            return node, deleted
        elif data > node.data:
            node.right, deleted = self._delete_recursive(node.right, data)
            return node, deleted
        else:
            # Node to delete found
            if node.left is None:
                return node.right, True
            elif node.right is None:
                return node.left, True
            else:
                # Node has two children
                min_node = self._find_min(node.right)
                node.data = min_node.data
                node.right, _ = self._delete_recursive(node.right, min_node.data)
                return node, True

    def _find_min(self, node: TreeNode) -> TreeNode:
        """Find the minimum value node in a subtree."""
        while node.left is not None:
            node = node.left
        return node

    def _find_max(self, node: TreeNode) -> TreeNode:
        """Find the maximum value node in a subtree."""
        while node.right is not None:
            node = node.right
        return node

    # Traversal Methods

    def inorder_traversal(self) -> Iterator[Any]:
        """In-order traversal (left, root, right)."""
        yield from self._inorder_recursive(self.root)

    def _inorder_recursive(self, node: Optional[TreeNode]) -> Iterator[Any]:
        """Helper for in-order traversal."""
        if node is not None:
            yield from self._inorder_recursive(node.left)
            yield node.data
            yield from self._inorder_recursive(node.right)

    def preorder_traversal(self) -> Iterator[Any]:
        """Pre-order traversal (root, left, right)."""
        yield from self._preorder_recursive(self.root)

    def _preorder_recursive(self, node: Optional[TreeNode]) -> Iterator[Any]:
        """Helper for pre-order traversal."""
        if node is not None:
            yield node.data
            yield from self._preorder_recursive(node.left)
            yield from self._preorder_recursive(node.right)

    def postorder_traversal(self) -> Iterator[Any]:
        """Post-order traversal (left, right, root)."""
        yield from self._postorder_recursive(self.root)

    def _postorder_recursive(self, node: Optional[TreeNode]) -> Iterator[Any]:
        """Helper for post-order traversal."""
        if node is not None:
            yield from self._postorder_recursive(node.left)
            yield from self._postorder_recursive(node.right)
            yield node.data

    def level_order_traversal(self) -> Iterator[Any]:
        """Level-order (breadth-first) traversal."""
        if self.root is None:
            return

        queue = deque([self.root])
        while queue:
            node = queue.popleft()
            yield node.data

            if node.left is not None:
                queue.append(node.left)
            if node.right is not None:
                queue.append(node.right)

    # Utility Methods

    def height(self) -> int:
        """Calculate the height of the tree."""
        return self._height_recursive(self.root)

    def _height_recursive(self, node: Optional[TreeNode]) -> int:
        """Helper for calculating height."""
        if node is None:
            return -1

        left_height = self._height_recursive(node.left)
        right_height = self._height_recursive(node.right)

        return 1 + max(left_height, right_height)

    def depth(self, data: Any) -> int:
        """Calculate the depth of a specific node."""
        return self._depth_recursive(self.root, data, 0)

    def _depth_recursive(
        self, node: Optional[TreeNode], data: Any, current_depth: int
    ) -> int:
        """Helper for calculating depth."""
        if node is None:
            return -1

        if node.data == data:
            return current_depth

        if data < node.data:
            return self._depth_recursive(node.left, data, current_depth + 1)
        else:
            return self._depth_recursive(node.right, data, current_depth + 1)

    def is_balanced(self) -> bool:
        """Check if the tree is height-balanced."""
        return self._is_balanced_recursive(self.root)[0]

    def _is_balanced_recursive(self, node: Optional[TreeNode]) -> tuple[bool, int]:
        """Helper for checking if tree is balanced."""
        if node is None:
            return True, -1

        left_balanced, left_height = self._is_balanced_recursive(node.left)
        right_balanced, right_height = self._is_balanced_recursive(node.right)

        balanced = (
            left_balanced and right_balanced and abs(left_height - right_height) <= 1
        )
        height = 1 + max(left_height, right_height)

        return balanced, height

    def count_nodes(self) -> int:
        """Count the total number of nodes."""
        return self.size

    def count_leaves(self) -> int:
        """Count the number of leaf nodes."""
        return self._count_leaves_recursive(self.root)

    def _count_leaves_recursive(self, node: Optional[TreeNode]) -> int:
        """Helper for counting leaf nodes."""
        if node is None:
            return 0

        if node.left is None and node.right is None:
            return 1

        return self._count_leaves_recursive(node.left) + self._count_leaves_recursive(
            node.right
        )

    def get_level_nodes(self, level: int) -> List[Any]:
        """Get all nodes at a specific level."""
        result = []
        self._get_level_nodes_recursive(self.root, level, 0, result)
        return result

    def _get_level_nodes_recursive(
        self,
        node: Optional[TreeNode],
        target_level: int,
        current_level: int,
        result: List[Any],
    ) -> None:
        """Helper for getting nodes at a specific level."""
        if node is None:
            return

        if current_level == target_level:
            result.append(node.data)
        else:
            self._get_level_nodes_recursive(
                node.left, target_level, current_level + 1, result
            )
            self._get_level_nodes_recursive(
                node.right, target_level, current_level + 1, result
            )

    def to_list(self) -> List[Any]:
        """Convert tree to sorted list (in-order traversal)."""
        return list(self.inorder_traversal())

    def to_json(self) -> str:
        """Convert tree to JSON representation."""
        return json.dumps(self._to_dict(self.root), indent=2)

    def _to_dict(self, node: Optional[TreeNode]) -> Optional[dict]:
        """Helper for JSON conversion."""
        if node is None:
            return None

        return {
            "data": node.data,
            "left": self._to_dict(node.left),
            "right": self._to_dict(node.right),
        }

    def pretty_print(self) -> None:
        """Print the tree in a visual format."""
        if self.root is None:
            print("Empty tree")
            return

        self._pretty_print_recursive(self.root, "", True)

    def _pretty_print_recursive(
        self, node: Optional[TreeNode], prefix: str, is_last: bool
    ) -> None:
        """Helper for pretty printing."""
        if node is not None:
            print(prefix + ("└── " if is_last else "├── ") + str(node.data))

            children = []
            if node.left is not None:
                children.append(node.left)
            if node.right is not None:
                children.append(node.right)

            for i, child in enumerate(children):
                is_last_child = i == len(children) - 1
                extension = "    " if is_last else "│   "
                self._pretty_print_recursive(child, prefix + extension, is_last_child)

    # Magic Methods

    def __len__(self) -> int:
        return self.size

    def __contains__(self, data: Any) -> bool:
        return self.search(data)

    def __bool__(self) -> bool:
        return self.root is not None

    def __iter__(self) -> Iterator[Any]:
        return self.inorder_traversal()

    def __str__(self) -> str:
        return str(list(self.inorder_traversal()))

    def __repr__(self) -> str:
        return f"BinaryTree({list(self.inorder_traversal())})"


def demonstrate_binary_tree():
    """Demonstrate various binary tree operations."""
    print("🌳 Binary Tree Implementation in Python")
    print("=" * 45)

    # Create and populate tree
    tree = BinaryTree()
    values = [50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45]

    print("\n📥 Inserting values:", values)
    for value in values:
        tree.insert(value)

    print(f"Tree size: {len(tree)}")
    print(f"Tree height: {tree.height()}")
    print(f"Is balanced: {tree.is_balanced()}")

    # Visual representation
    print("\n🎨 Tree Structure:")
    tree.pretty_print()

    # Traversals
    print("\n🚶 Tree Traversals:")
    print(f"In-order:    {list(tree.inorder_traversal())}")
    print(f"Pre-order:   {list(tree.preorder_traversal())}")
    print(f"Post-order:  {list(tree.postorder_traversal())}")
    print(f"Level-order: {list(tree.level_order_traversal())}")

    # Search operations
    print("\n🔍 Search Operations:")
    search_values = [25, 75, 50, 100]
    for value in search_values:
        found = value in tree
        depth = tree.depth(value) if found else -1
        print(f"Search {value}: {'Found' if found else 'Not found'} (depth: {depth})")

    # Level operations
    print("\n📏 Level Operations:")
    for level in range(tree.height() + 1):
        nodes = tree.get_level_nodes(level)
        print(f"Level {level}: {nodes}")

    # Statistics
    print(f"\n📊 Tree Statistics:")
    print(f"Total nodes: {tree.count_nodes()}")
    print(f"Leaf nodes: {tree.count_leaves()}")
    print(f"Height: {tree.height()}")

    # Deletion
    print(f"\n🗑️ Deletion Operations:")
    delete_values = [10, 30, 50]
    for value in delete_values:
        print(f"Deleting {value}...")
        deleted = tree.delete(value)
        print(f"Success: {deleted}, New tree: {tree}")
        if deleted:
            print("Updated tree structure:")
            tree.pretty_print()

    # JSON representation
    print(f"\n📄 JSON Representation:")
    print(tree.to_json())


def performance_test():
    """Test performance with larger datasets."""
    import random
    import time

    print("\n⚡ Performance Testing")
    print("=" * 25)

    sizes = [1000, 5000, 10000]

    for size in sizes:
        print(f"\nTesting with {size} elements:")

        # Generate random data
        data = list(range(size))
        random.shuffle(data)

        tree = BinaryTree()

        # Test insertion
        start_time = time.time()
        for value in data:
            tree.insert(value)
        insertion_time = time.time() - start_time

        # Test search
        search_data = random.sample(data, min(100, len(data)))
        start_time = time.time()
        for value in search_data:
            tree.search(value)
        search_time = time.time() - start_time

        # Test traversal
        start_time = time.time()
        list(tree.inorder_traversal())
        traversal_time = time.time() - start_time

        print(f"  Insertion:  {insertion_time:.4f}s")
        print(f"  Search:     {search_time:.4f}s")
        print(f"  Traversal:  {traversal_time:.4f}s")
        print(f"  Height:     {tree.height()}")
        print(f"  Balanced:   {tree.is_balanced()}")


if __name__ == "__main__":
    demonstrate_binary_tree()
    performance_test()

    print("\n✨ Binary Tree demonstration complete!")

File: data-structures/binary_tree.js
/**
 * Binary Tree Implementation in JavaScript
 * 
 * Time Complexity:
 * - Insert: O(log n) average, O(n) worst case
 * - Search: O(log n) average, O(n) worst case
 * - Delete: O(log n) average, O(n) worst case
 * 
 * Space Complexity: O(n) for storage, O(log n) for recursion stack
 * 
 * JavaScript features:
 * - ES6+ syntax
 * - Generator functions
 * - Classes and modules
 * - Async/await support
 */

class TreeNode {
    constructor(data) {
        this.data = data;
        this.left = null;
        this.right = null;
    }
    
    toString() {
        return this.data.toString();
    }
}

class BinaryTree {
    constructor() {
        this.root = null;
        this.size = 0;
    }
    
    // Core Operations
    
    insert(data) {
        if (this.root === null) {
            this.root = new TreeNode(data);
        } else {
            this._insertRecursive(this.root, data);
        }
        this.size++;
        return this;
    }
    
    _insertRecursive(node, data) {
        if (data < node.data) {
            if (node.left === null) {
                node.left = new TreeNode(data);
            } else {
                this._insertRecursive(node.left, data);
            }
        } else {
            if (node.right === null) {
                node.right = new TreeNode(data);
            } else {
                this._insertRecursive(node.right, data);
            }
        }
    }
    
    search(data) {
        return this._searchRecursive(this.root, data);
    }
    
    _searchRecursive(node, data) {
        if (node === null) {
            return false;
        }
        
        if (data === node.data) {
            return true;
        } else if (data < node.data) {
            return this._searchRecursive(node.left, data);
        } else {
            return this._searchRecursive(node.right, data);
        }
    }
    
    delete(data) {
        if (this.root === null) {
            return false;
        }
        
        const result = this._deleteRecursive(this.root, data);
        this.root = result.node;
        
        if (result.deleted) {
            this.size--;
        }
        
        return result.deleted;
    }
    
    _deleteRecursive(node, data) {
        if (node === null) {
            return { node: null, deleted: false };
        }
        
        if (data < node.data) {
            const result = this._deleteRecursive(node.left, data);
            node.left = result.node;
            return { node, deleted: result.deleted };
        } else if (data > node.data) {
            const result = this._deleteRecursive(node.right, data);
            node.right = result.node;
            return { node, deleted: result.deleted };
        } else {
            // Node to delete found
            if (node.left === null) {
                return { node: node.right, deleted: true };
            } else if (node.right === null) {
                return { node: node.left, deleted: true };
            } else {
                // Node has two children
                const minNode = this._findMin(node.right);
                node.data = minNode.data;
                const result = this._deleteRecursive(node.right, minNode.data);
                node.right = result.node;
                return { node, deleted: true };
            }
        }
    }
    
    _findMin(node) {
        while (node.left !== null) {
            node = node.left;
        }
        return node;
    }
    
    _findMax(node) {
        while (node.right !== null) {
            node = node.right;
        }
        return node;
    }
    
    // Traversal Methods (Using Generators)
    
    *inorderTraversal(node = this.root) {
        if (node !== null) {
            yield* this.inorderTraversal(node.left);
            yield node.data;
            yield* this.inorderTraversal(node.right);
        }
    }
    
    *preorderTraversal(node = this.root) {
        if (node !== null) {
            yield node.data;
            yield* this.preorderTraversal(node.left);
            yield* this.preorderTraversal(node.right);
        }
    }
    
    *postorderTraversal(node = this.root) {
        if (node !== null) {
            yield* this.postorderTraversal(node.left);
            yield* this.postorderTraversal(node.right);
            yield node.data;
        }
    }
    
    *levelOrderTraversal() {
        if (this.root === null) return;
        
        const queue = [this.root];
        while (queue.length > 0) {
            const node = queue.shift();
            yield node.data;
            
            if (node.left !== null) {
                queue.push(node.left);
            }
            if (node.right !== null) {
                queue.push(node.right);
            }
        }
    }
    
    // Utility Methods
    
    height(node = this.root) {
        if (node === null) {
            return -1;
        }
        
        const leftHeight = this.height(node.left);
        const rightHeight = this.height(node.right);
        
        return 1 + Math.max(leftHeight, rightHeight);
    }
    
    depth(data, node = this.root, currentDepth = 0) {
        if (node === null) {
            return -1;
        }
        
        if (node.data === data) {
            return currentDepth;
        }
        
        if (data < node.data) {
            return this.depth(data, node.left, currentDepth + 1);
        } else {
            return this.depth(data, node.right, currentDepth + 1);
        }
    }
    
    isBalanced(node = this.root) {
        const result = this._isBalancedRecursive(node);
        return result.balanced;
    }
    
    _isBalancedRecursive(node) {
        if (node === null) {
            return { balanced: true, height: -1 };
        }
        
        const leftResult = this._isBalancedRecursive(node.left);
        const rightResult = this._isBalancedRecursive(node.right);
        
        const balanced = leftResult.balanced && 
                        rightResult.balanced && 
                        Math.abs(leftResult.height - rightResult.height) <= 1;
        
        const height = 1 + Math.max(leftResult.height, rightResult.height);
        
        return { balanced, height };
    }
    
    countNodes() {
        return this.size;
    }
    
    countLeaves(node = this.root) {
        if (node === null) {
            return 0;
        }
        
        if (node.left === null && node.right === null) {
            return 1;
        }
        
        return this.countLeaves(node.left) + this.countLeaves(node.right);
    }
    
    getLevelNodes(targetLevel, node = this.root, currentLevel = 0) {
        if (node === null) {
            return [];
        }
        
        if (currentLevel === targetLevel) {
            return [node.data];
        }
        
        return [
            ...this.getLevelNodes(targetLevel, node.left, currentLevel + 1),
            ...this.getLevelNodes(targetLevel, node.right, currentLevel + 1)
        ];
    }
    
    toArray() {
        return [...this.inorderTraversal()];
    }
    
    toJSON() {
        return JSON.stringify(this._nodeToObject(this.root), null, 2);
    }
    
    _nodeToObject(node) {
        if (node === null) {
            return null;
        }
        
        return {
            data: node.data,
            left: this._nodeToObject(node.left),
            right: this._nodeToObject(node.right)
        };
    }
    
    prettyPrint(node = this.root, prefix = '', isLast = true) {
        if (this.root === null) {
            console.log('Empty tree');
            return;
        }
        
        if (node !== null) {
            console.log(prefix + (isLast ? '└── ' : '├── ') + node.data);
            
            const children = [];
            if (node.left !== null) children.push(node.left);
            if (node.right !== null) children.push(node.right);
            
            children.forEach((child, index) => {
                const isLastChild = index === children.length - 1;
                const extension = isLast ? '    ' : '│   ';
                this.prettyPrint(child, prefix + extension, isLastChild);
            });
        }
    }
    
    // Iterator Support
    
    [Symbol.iterator]() {
        return this.inorderTraversal();
    }
    
    // Conversion Methods
    
    toString() {
        return JSON.stringify([...this.inorderTraversal()]);
    }
    
    // Validation
    
    isValidBST(node = this.root, min = -Infinity, max = Infinity) {
        if (node === null) {
            return true;
        }
        
        if (node.data <= min || node.data >= max) {
            return false;
        }
        
        return this.isValidBST(node.left, min, node.data) &&
               this.isValidBST(node.right, node.data, max);
    }
    
    // Advanced Operations
    
    findLCA(data1, data2) {
        return this._findLCARecursive(this.root, data1, data2);
    }
    
    _findLCARecursive(node, data1, data2) {
        if (node === null) {
            return null;
        }
        
        if (data1 < node.data && data2 < node.data) {
            return this._findLCARecursive(node.left, data1, data2);
        } else if (data1 > node.data && data2 > node.data) {
            return this._findLCARecursive(node.right, data1, data2);
        } else {
            return node.data;
        }
    }
    
    getPathToNode(data) {
        const path = [];
        this._getPathRecursive(this.root, data, path);
        return path;
    }
    
    _getPathRecursive(node, data, path) {
        if (node === null) {
            return false;
        }
        
        path.push(node.data);
        
        if (node.data === data) {
            return true;
        }
        
        if (data < node.data && this._getPathRecursive(node.left, data, path)) {
            return true;
        }
        
        if (data > node.data && this._getPathRecursive(node.right, data, path)) {
            return true;
        }
        
        path.pop();
        return false;
    }
    
    // Static Factory Methods
    
    static fromArray(array) {
        const tree = new BinaryTree();
        array.forEach(value => tree.insert(value));
        return tree;
    }
    
    static fromSortedArray(sortedArray) {
        const tree = new BinaryTree();
        tree.root = BinaryTree._buildBalancedFromSorted(sortedArray, 0, sortedArray.length - 1);
        tree.size = sortedArray.length;
        return tree;
    }
    
    static _buildBalancedFromSorted(array, start, end) {
        if (start > end) {
            return null;
        }
        
        const mid = Math.floor((start + end) / 2);
        const node = new TreeNode(array[mid]);
        
        node.left = BinaryTree._buildBalancedFromSorted(array, start, mid - 1);
        node.right = BinaryTree._buildBalancedFromSorted(array, mid + 1, end);
        
        return node;
    }
}

// Utility Functions

function demonstrateBinaryTree() {
    console.log('🌳 Binary Tree Implementation in JavaScript');
    console.log('='.repeat(45));
    
    // Create and populate tree
    const tree = new BinaryTree();
    const values = [50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45];
    
    console.log('\n📥 Inserting values:', values);
    values.forEach(value => tree.insert(value));
    
    console.log(`Tree size: ${tree.countNodes()}`);
    console.log(`Tree height: ${tree.height()}`);
    console.log(`Is balanced: ${tree.isBalanced()}`);
    console.log(`Is valid BST: ${tree.isValidBST()}`);
    
    // Visual representation
    console.log('\n🎨 Tree Structure:');
    tree.prettyPrint();
    
    // Traversals
    console.log('\n🚶 Tree Traversals:');
    console.log(`In-order:    [${[...tree.inorderTraversal()].join(', ')}]`);
    console.log(`Pre-order:   [${[...tree.preorderTraversal()].join(', ')}]`);
    console.log(`Post-order:  [${[...tree.postorderTraversal()].join(', ')}]`);
    console.log(`Level-order: [${[...tree.levelOrderTraversal()].join(', ')}]`);
    
    // Search operations
    console.log('\n🔍 Search Operations:');
    const searchValues = [25, 75, 50, 100];
    searchValues.forEach(value => {
        const found = tree.search(value);
        const depth = found ? tree.depth(value) : -1;
        console.log(`Search ${value}: ${found ? 'Found' : 'Not found'} (depth: ${depth})`);
    });
    
    // Level operations
    console.log('\n📏 Level Operations:');
    for (let level = 0; level <= tree.height(); level++) {
        const nodes = tree.getLevelNodes(level);
        console.log(`Level ${level}: [${nodes.join(', ')}]`);
    }
    
    // Advanced operations
    console.log('\n🔍 Advanced Operations:');
    console.log(`LCA of 25 and 35: ${tree.findLCA(25, 35)}`);
    console.log(`Path to 25: [${tree.getPathToNode(25).join(' → ')}]`);
    console.log(`Path to 80: [${tree.getPathToNode(80).join(' → ')}]`);
    
    // Statistics
    console.log(`\n📊 Tree Statistics:`);
    console.log(`Total nodes: ${tree.countNodes()}`);
    console.log(`Leaf nodes: ${tree.countLeaves()}`);
    console.log(`Height: ${tree.height()}`);
    
    // Deletion
    console.log(`\n🗑️ Deletion Operations:`);
    const deleteValues = [10, 30, 50];
    deleteValues.forEach(value => {
        console.log(`Deleting ${value}...`);
        const deleted = tree.delete(value);
        console.log(`Success: ${deleted}, New tree: ${tree}`);
        if (deleted) {
            console.log('Updated tree structure:');
            tree.prettyPrint();
        }
    });
}

function performanceTest() {
    console.log('\n⚡ Performance Testing');
    console.log('='.repeat(25));
    
    const sizes = [1000, 5000, 10000];
    
    sizes.forEach(size => {
        console.log(`\nTesting with ${size} elements:`);
        
        // Generate random data
        const data = Array.from({ length: size }, (_, i) => i);
        for (let i = data.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [data[i], data[j]] = [data[j], data[i]];
        }
        
        const tree = new BinaryTree();
        
        // Test insertion
        const insertStart = performance.now();
        data.forEach(value => tree.insert(value));
        const insertTime = performance.now() - insertStart;
        
        // Test search
        const searchData = data.slice(0, Math.min(100, data.length));
        const searchStart = performance.now();
        searchData.forEach(value => tree.search(value));
        const searchTime = performance.now() - searchStart;
        
        // Test traversal
        const traversalStart = performance.now();
        [...tree.inorderTraversal()];
        const traversalTime = performance.now() - traversalStart;
        
        console.log(`  Insertion:  ${insertTime.toFixed(2)}ms`);
        console.log(`  Search:     ${searchTime.toFixed(2)}ms`);
        console.log(`  Traversal:  ${traversalTime.toFixed(2)}ms`);
        console.log(`  Height:     ${tree.height()}`);
        console.log(`  Balanced:   ${tree.isBalanced()}`);
    });
}

function testFactoryMethods() {
    console.log('\n🏭 Factory Methods Testing');
    console.log('='.repeat(30));
    
    // Test fromArray
    const array = [3, 1, 4, 1, 5, 9, 2, 6, 5];
    const treeFromArray = BinaryTree.fromArray(array);
    console.log(`From array [${array.join(', ')}]:`);
    console.log(`Result: [${[...treeFromArray].join(', ')}]`);
    console.log(`Height: ${treeFromArray.height()}, Balanced: ${treeFromArray.isBalanced()}`);
    
    // Test fromSortedArray (creates balanced tree)
    const sortedArray = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    const balancedTree = BinaryTree.fromSortedArray(sortedArray);
    console.log(`\nFrom sorted array [${sortedArray.join(', ')}]:`);
    console.log(`Result: [${[...balancedTree].join(', ')}]`);
    console.log(`Height: ${balancedTree.height()}, Balanced: ${balancedTree.isBalanced()}`);
    
    console.log('\nBalanced tree structure:');
    balancedTree.prettyPrint();
}

// Module exports for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        TreeNode,
        BinaryTree,
        demonstrateBinaryTree,
        performanceTest,
        testFactoryMethods
    };
}

// Main execution
if (typeof window === 'undefined' && typeof global !== 'undefined') {
    // Node.js environment
    demonstrateBinaryTree();
    performanceTest();
    testFactoryMethods();
    
    console.log('\n✨ Binary Tree demonstration complete!');
}
